### PR TITLE
Feature/pre mtu chain writing

### PR DESCRIPTION
**Pull request by jthomas43**
defers creating packets until they are ready to be flushed. this is to enable the packet size (MTU) to be adjusted on the fly.

### CODE REVIEW

File: include/udx.h [hunk 1]
```diff
@@ -267,7 +275,7 @@ struct udx_socket_send_s {
 };
 
 struct udx_stream_write_s {
-  uint32_t packets;
+  int64_t bytes;
```
Comment (kasperisager) [inline include/udx.h:278]: Any reason for this being signed? Why not `size_t`?

File: include/udx.h [hunk 2]
```diff
@@ -82,6 +82,13 @@ typedef struct udx_socket_send_s udx_socket_send_t;
 typedef struct udx_stream_write_s udx_stream_write_t;
 typedef struct udx_stream_send_s udx_stream_send_t;
 
+/* todo: should this be merged into the stream_write_t object? */
+typedef struct {
+  uv_buf_t buf;
+  udx_stream_write_t *write;
+  unsigned is_write_end : 1;
```
Comment (kasperisager) [inline include/udx.h:89]:
```suggestion
  bool is_write_end;
```

C bitfields are really wonky and even at best the struct will be appropriately padded anyway.

File: include/udx.h [hunk 3]
```diff
@@ -82,6 +82,13 @@ typedef struct udx_socket_send_s udx_socket_send_t;
 typedef struct udx_stream_write_s udx_stream_write_t;
 typedef struct udx_stream_send_s udx_stream_send_t;
 
+/* todo: should this be merged into the stream_write_t object? */
```
Comment (kasperisager) [inline include/udx.h:85]: If there's always a 1:1 correspondence between them, I'd merge them.

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: include/udx.h (hunk 1 above)
Comment (jthomas43) [reply include/udx.h]: changed

File: include/udx.h (hunk 2 above)
Comment (jthomas43) [reply include/udx.h]: changed

File: include/udx.h (hunk 3 above)
Comment (jthomas43) [reply include/udx.h]: for now the stream_write call only accepts a single buffer, but the plan is still to allow an multiple bufs to be passed to a single write, so it makes sense to keep the separate object. I removed the todo.

File: src/fifo.c [hunk 4]
```diff
@@ -32,6 +32,13 @@ udx__fifo_shift (udx_fifo_t *f) {
   return *b;
 }
 
+void *
+udx__fifo_first (udx_fifo_t *f) {
```
Comment (mafintosh) [inline src/fifo.c:36]: can we name this peak? we do that elsewhere

Comment (mafintosh) [reply src/fifo.c:36]: peek

File: src/udx.c [hunk 5]
```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
...
+    udx_write_buffer_t *wbuf = udx__fifo_first(&stream->write_buffer_queue);
+    assert(wbuf != NULL);
+    uv_buf_t *buf = &wbuf->buf;
+    assert(buf != NULL);
+
+    udx_stream_write_t *w = wbuf->write;
+
+    int header_flag = wbuf->is_write_end ? UDX_HEADER_END : UDX_HEADER_DATA;
+
+    int len = get_window_bytes(stream);
```
Comment (mafintosh) [inline src/udx.c:643]: uint32_t ya?

File: src/udx.c [hunk 6]
```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
...
+    if (buf->len < len) len = buf->len;
+    if (stream->mtu < len) len = stream->mtu;
+
+    if (stream->mtu > len && buf->len > len && stream->pkts_inflight > 0) {
```
Comment (mafintosh) [inline src/udx.c:647]: when is this true given the clamping above?

File: src/udx.c [hunk 7]
```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
...
   }
 
   if (sent < 0) return sent;
   return 0;
 }
 
+static inline uint32_t
+get_window_bytes (udx_stream_t *stream) {
+  // todo: receive window, then return lesser of cwnd, rwnd
+  // todo:
```
Comment (mafintosh) [inline src/udx.c:612]: empty todo

File: src/udx.c [hunk 8]
```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
...
+
+  while (get_window_bytes(stream) > 0 && stream->write_buffer_queue.len > 0) {
+    udx_write_buffer_t *wbuf = udx__fifo_first(&stream->write_buffer_queue);
+    assert(wbuf != NULL);
+    uv_buf_t *buf = &wbuf->buf;
+    assert(buf != NULL);
+
+    udx_stream_write_t *w = wbuf->write;
+
+    int header_flag = wbuf->is_write_end ? UDX_HEADER_END : UDX_HEADER_DATA;
```
Comment (mafintosh) [inline src/udx.c:641]: if the packet has data AND is_write_end if needs both bits set (END | DATA)

File: src/udx.c [hunk 9]
```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
...
+
+    pkt->fifo_gc = udx__fifo_push(&stream->socket->send_queue, pkt);
+
+    if (buf->len == 0) {
+      udx__fifo_shift(&stream->write_buffer_queue);
+      free(wbuf);
+    }
+
+    int rc = update_poll(stream->socket);
+    // maybe just assert? can only return EEXIST in a narrow circumstance?
```
Comment (mafintosh) [inline src/udx.c:681]: perfectly happy to update all our update_poll error handling to asserts

File: src/udx.c [hunk 10]
```diff
@@ -584,13 +598,98 @@ flush_waiting_packets (udx_stream_t *stream) {
...
+
+    if (rc < 0) {
+      return rc;
+    }
+  }
+
+  // debug_print_outgoing(stream);
```
Comment (mafintosh) [inline src/udx.c:688]: remove thhe log

File: src/udx.c [hunk 11]
```diff
@@ -705,6 +805,7 @@ rack_detect_loss (udx_stream_t *stream) {
 
   if (resending) {
     if (stream->recovery == 0) {
+      // debug_print_outgoing(stream);
```
Comment (mafintosh) [inline src/udx.c:808]: remove this

File: src/udx.c [hunk 12]
```diff
@@ -1611,7 +1715,7 @@ udx_stream_check_timeouts (udx_stream_t *handle) {
     return 0;
   }
 
-  if (handle->remote_acked == handle->seq) {
+  if (handle->remote_acked == handle->seq && handle->write_buffer_queue.len == 0) {
```
Comment (mafintosh) [inline src/udx.c:1718]: nice

File: src/udx.c (hunk 6 above)
Comment (jthomas43) [reply src/udx.c:647]: in the last iteration of the loop, if neither of the clamps are triggered, e.g. remaining window size is less than the remaining buffer, and remaining window size is less than the stream mtu value.

the conditional there bails out of sending the small packet at the end of the flight, unless there are zero packets in flight and we need to send a small packet to make progress (e.g. the whole write is only one packet). It's an optimization kind of like TCP Cork.

Comment (mafintosh) [reply src/udx.c:642]: Does this delay sending a packet? Cause we rely on packets not being queued unless they have to (ie no naggle algo)

**Pull request merged**